File: Ejercicios/State.java
package Ejercicios;

interface EstadoTarea {
    EstadoTarea comenzar();
    EstadoTarea completar();
}


class Pendiente implements EstadoTarea {
    public EstadoTarea comenzar() {
        System.out.println("Tarea en progreso...");
        return new EnProgreso();
    }

    public EstadoTarea completar() {
        System.out.println("No se puede completar una tarea pendiente");
        return this;
    }
}


class EnProgreso implements EstadoTarea {
    public EstadoTarea comenzar() {
        System.out.println("La tarea ya está en progreso");
        return this;
    }

    public EstadoTarea completar() {
        System.out.println("Tarea completada!");
        return new Completada();
    }
}


class Completada implements EstadoTarea {
    public EstadoTarea comenzar() {
        System.out.println("La tarea ya fue completada");
        return this;
    }

    public EstadoTarea completar() {
        System.out.println("La tarea ya está completada");
        return this;
    }
}


class Tarea {
    private String nombre ;
    private EstadoTarea estado;

    public Tarea(String nombre) {
        this.nombre = nombre;
        this.estado = new Pendiente();
    }

    public void comenzar() {
        estado = estado.comenzar();
    }

    public void completar() {
        estado = estado.completar();
    }
}


public class State {
    public static void main(String[] args) {
        Tarea tarea1 = new Tarea("Desarrollar módulo de login");

        tarea1.completar(); 
        tarea1.comenzar();  
        tarea1.comenzar();  
        tarea1.completar(); 
        tarea1.completar();
    }
}

File: Ejercicios/state.py
from abc import ABC, abstractmethod


class EstadoTarea(ABC):
    @abstractmethod
    def comenzar(self):
        pass

    @abstractmethod
    def completar(self):
        pass


class Pendiente(EstadoTarea):
    def comenzar(self):
        print("Tarea en progreso...")
        return EnProgreso()
    
    def completar(self):
        print("No se puede completar una tarea pendiente")
        return self


class EnProgreso(EstadoTarea):
    def comenzar(self):
        print("La tarea ya está en progreso")
        return self
    
    def completar(self):
        print("Tarea completada!")
        return Completada()

class Completada(EstadoTarea):
    def comenzar(self):
        print("La tarea ya fue completada")
        return self
    
    def completar(self):
        print("La tarea ya está completada")
        return self


class Tarea:
    def __init__(self, nombre):
        self.nombre = nombre
        self.estado = Pendiente()
    
    def comenzar(self):
        self.estado = self.estado.comenzar()
    
    def completar(self):
        self.estado = self.estado.completar()

tarea1 = Tarea("Desarrollar módulo de login")

tarea1.completar()    
tarea1.comenzar()     
tarea1.comenzar()     
tarea1.completar()    
tarea1.completar()    
